File: 250314Shenfeng/250328multiplefileasgroup.py
import os
import pandas as pd

# Folder containing the CSV files
folder_path = '/Users/w/Developer/250306 KYK Datamining/250314Shenfeng/datasets'

# Get a list of all CSV files in the folder
csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]

# Dictionary to store information about each file
file_info = {}

# Loop through each CSV file and collect information
for file in csv_files:
    file_path = os.path.join(folder_path, file)
    try:
        df = pd.read_csv(file_path)
        file_info[file] = {
            'columns': tuple(df.columns),  # Use tuple to make it hashable
            'dtypes': tuple(df.dtypes.items()),  # Use tuple of dtype items for comparison
            'shape': df.shape
        }
    except Exception as e:
        print(f"Error reading {file}: {e}")

# Group files by structure (columns and data types)
grouped_files = {}
for file, info in file_info.items():
    key = (info['columns'], info['dtypes'])  # Group by columns and data types
    if key not in grouped_files:
        grouped_files[key] = []
    grouped_files[key].append(file)

# Combine files in each group
output_folder = '/Users/w/Developer/250306 KYK Datamining/250314Shenfeng/combined'
os.makedirs(output_folder, exist_ok=True)

for i, (key, files) in enumerate(grouped_files.items(), 1):
    print(f"Combining files in Group {i}: {files}")
    combined_data = pd.DataFrame()
    for file in files:
        file_path = os.path.join(folder_path, file)
        df = pd.read_csv(file_path)
        combined_data = pd.concat([combined_data, df], ignore_index=True)
    
    # Save the combined data as a new CSV file
    output_file = os.path.join(output_folder, f'group_{i}_combined.csv')
    combined_data.to_csv(output_file, index=False)
    print(f"Group {i} combined file saved as: {output_file}")

File: 250314Shenfeng/250328multiplefileasgroup.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <filesystem>
#include <sstream>
#include <Eigen/Dense>

namespace fs = std::filesystem;

struct CSVInfo {
    std::vector<std::string> columns;
    std::vector<std::string> dtypes;
    std::pair<int,int> shape;
};

CSVInfo analyze_csv(const std::string& file_path) {
    std::ifstream file(file_path);
    std::string line;
    CSVInfo info;
    if (std::getline(file, line)) {
        std::stringstream ss(line);
        std::string col;
        while (std::getline(ss, col, ',')) info.columns.push_back(col);
    }
    int row_count = 0;
    while (std::getline(file, line)) ++row_count;
    info.shape = {row_count, (int)info.columns.size()};
    info.dtypes = std::vector<std::string>(info.columns.size(), "string/double"); // C++ does not infer dtypes
    return info;
}

void combine_csvs(const std::vector<std::string>& files, const std::string& folder_path, const std::string& output_file) {
    std::ofstream out(output_file);
    bool header_written = false;
    for (const auto& file : files) {
        std::ifstream in(folder_path + "/" + file);
        std::string line;
        bool first_line = true;
        while (std::getline(in, line)) {
            if (first_line) {
                if (!header_written) {
                    out << line << "\n";
                    header_written = true;
                }
                first_line = false;
            } else {
                out << line << "\n";
            }
        }
    }
    out.close();
}

int main() {
    std::string folder_path = "250314Shenfeng/datasets";
    std::string output_folder = "250314Shenfeng/combined";
    fs::create_directories(output_folder);
    std::map<std::string, CSVInfo> file_info;
    for (const auto& entry : fs::directory_iterator(folder_path)) {
        if (entry.path().extension() == ".csv") {
            try {
                file_info[entry.path().filename()] = analyze_csv(entry.path().string());
            } catch (...) {
                std::cout << "Error reading " << entry.path().filename() << std::endl;
            }
        }
    }
    // Group files by structure (columns and dtypes)
    std::map<std::pair<std::vector<std::string>, std::vector<std::string>>, std::vector<std::string>> grouped_files;
    for (const auto& [file, info] : file_info) {
        auto key = std::make_pair(info.columns, info.dtypes);
        grouped_files[key].push_back(file);
    }
    int group_num = 1;
    for (const auto& [key, files] : grouped_files) {
        std::cout << "Combining files in Group " << group_num << ": ";
        for (auto& f : files) std::cout << f << " ";
        std::cout << std::endl;
        std::string output_file = output_folder + "/group_" + std::to_string(group_num) + "_combined.csv";
        combine_csvs(files, folder_path, output_file);
        std::cout << "Group " << group_num << " combined file saved as: " << output_file << std::endl;
        ++group_num;
    }
    return 0;
}
